packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx

```diff
@@ -382,84 +457,349 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
       id={id}
     >
       <div className='iui-input-with-icon'>
-        <Popover
-          placement='bottom-start'
-          visible={isOpen}
-          onClickOutside={(_, { target }) => {
-            if (!toggleButtonRef.current?.contains(target as Element)) {
-              setIsOpen(false);
+        {children}
+        {endIcon}
+      </div>
+    </InputContainer>
+  );
+};
+ComboBoxInputContainer.displayName = 'ComboBoxInputContainer';
+
+const ComboBoxEndIcon = React.forwardRef(
+  (
+    props: React.ComponentPropsWithoutRef<'span'>,
+    forwardedRef: React.RefObject<HTMLSpanElement>,
+  ) => {
+    const { className, children, onClick: onClickProp, ...rest } = props;
+    const dispatch = useSafeContext(ComboBoxActionContext);
+    const { toggleButtonRef } = useSafeContext(ComboBoxRefsContext);
+    const refs = useMergedRefs(toggleButtonRef, forwardedRef);
+
+    return (
+      <span
+        ref={refs}
+        className={cx('iui-end-icon', className)}
+        onClick={(e) => {
+          onClickProp?.(e);
+          if (!e.isDefaultPrevented()) {
+            dispatch(['toggle']);
+          }
+        }}
+        {...rest}
+      >
+        {children ?? <SvgCaretDownSmall aria-hidden />}
+      </span>
+    );
+  },
+);
+ComboBoxEndIcon.displayName = 'ComboBoxEndIcon';
+
+const ComboBoxInput = React.forwardRef(
+  (props: InputProps, forwardedRef: React.RefObject<HTMLInputElement>) => {
+    const { onKeyDown: onKeyDownProp, onFocus: onFocusProp, ...rest } = props;
+
+    const { isOpen, id, focusedIndex } = useSafeContext(ComboBoxStateContext);
+    const dispatch = useSafeContext(ComboBoxActionContext);
+    const { inputRef, menuRef, optionsRef } =
+      useSafeContext(ComboBoxRefsContext);
+    const refs = useMergedRefs(inputRef, forwardedRef);
+
+    const focusedIndexRef = React.useRef(focusedIndex ?? -1);
+    React.useEffect(() => {
+      focusedIndexRef.current = focusedIndex ?? -1;
+    }, [focusedIndex]);
+
+    const getIdFromIndex = (index: number) => {
+      return (
+        menuRef.current?.querySelector(`[data-iui-index="${index}"]`)?.id ?? ''
+      );
+    };
+
+    const handleKeyDown = React.useCallback(
+      (event: React.KeyboardEvent<HTMLInputElement>) => {
+        onKeyDownProp?.(event);
+        const length = optionsRef.current.length ?? 0;
+
+        switch (event.key) {
+          case 'ArrowDown': {
```

Comment:
hm...whats not nice now, is that if you hold arrow, nothing happens.. You dont see list going down anymore..

Reply:
It is going down for me on holding the key. But I for me it is lagging. Opened deployed original version and it is navigating smoothly.

Reply:
Yeah, cause other one for me did not even load (disabled one)

Reply:
Holding down arrow works for me

Reply:
It still does not work for me... :/